### PR TITLE
ClausalBitvectorProof

### PR DESCRIPTION
**Pull request by alex-ozdemir**
## Summary

Created a class, `ClausalBitvectorProof`, which represents a bitvector
proof of UNSAT using an underlying clausal technique (DRAT, LRAT, etc)

It fits into the `BitvectorProof` class hierarchy like this:

```
              BitvectorProof
              /            \
             /              \
ClausalBitvectorProof  ResolutionBitvectorProof
```

## Why this is so nasty

This is probably going to be the nastiest diff in the DRAT/LRAT/ER sequence.

It is a painful one because all of the following BV subsystems
referenced ResolutionBitvectorProof (subsequently RBVP) or
BitvectorProof (subsequently BVP):
   * CnfStream
   * SatSolver (specifically the BvSatSolver)
   * CnfProof
   * TheoryProof
   * TheoryBV
   * Both bitblasters

And in particular, ResolutionBitvectorProof, the CnfStream, and the
SatSolvers were tightly coupled.

This means that references to and interactions with (R)BVP were
pervasive, and all of those references needed to be examined to figure out
which member of the new hierarchy they should refer to.

## Outline of changes

The change summary:
  * Create a subclass of BVP, called ClausalBitvectorProof, which has
    most methods stubbed out.
  * Make a some modifications to BVP and ResolutionBitvectorProof as the
    natural division of labor between the different classes becomes
    clear.
  * Go through all the components in the first list and try to figure
    out which kind of BVP they should **actually** be interacting with,
    and how. Make tweaks accordingly.
  * Add a hook from CryptoMinisat which pipes the produced DRAT proof
    into the new ClausalBitvectorProof.
  * Add a debug statement to ClausalBitvectorProof which parses and
    prints that DRAT proof, for testing purposes.

## Testing

Test:
  * `make check` to verify that we didn't break any old stuff, including
    lazy BB, and eager BB when using bvminisat.
  * `cvc4 --dump-proofs --bv-sat-solver=cryptominisat --bitblast=eager
  -d bv::clausal test/regress/regress0/bv/ackermann2.smt2`, and see that
     1. It crashed with "Unimplemented"
     2. Right before that it prints out the (textual) DRAT proof.

## Review
I assigned review to Andres, since he reviewed the BVP/RBVP class split, and this is a further articulation of that class hierarchy. However, if you think it would be better to have one the the BV developers review, feel free to re-assign.

### CODE REVIEW

File: src/proof/clausal_bitvector_proof.cpp [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+                                         prop::SatVariable falseVar)
+{
+  Assert(d_cnfProof == nullptr);
+  d_cnfProof.reset(new LFSCCnfProof(cnfStream, cnf, "bb"));
+
+  // Create a clause which forces the true variable to be true, and register it
+  int trueClauseId = ClauseId(ProofManager::currentPM()->nextId());
+  // with the CNF proof
+  d_cnfProof->registerTrueUnitClause(trueClauseId);
+  // and with (this) bitvector proof
```
Comment (4tXJ7f) [inline src/proof/clausal_bitvector_proof.cpp]:
```suggestion
  // and with (this) bit-vector proof
```

File: src/proof/clausal_bitvector_proof.cpp [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+  int falseClauseId = ClauseId(ProofManager::currentPM()->nextId());
+  d_cnfProof->registerFalseUnitClause(falseClauseId);
+  c[0] = prop::SatLiteral(falseVar, true);
+  registerUsedClause(falseClauseId, c);
+}
+
+void ClausalBitVectorProof::registerUsedClause(ClauseId id,
+                                               prop::SatClause& clause)
+{
+  d_usedClauses.push_back(std::make_pair(
```
Comment (4tXJ7f) [inline src/proof/clausal_bitvector_proof.cpp]: Suggestion: could use `emplace_back()` instead.

File: src/proof/bitvector_proof.h [hunk 3]
```diff
@@ -131,6 +137,28 @@ class BitVectorProof : public TheoryProof
...
+                                        std::ostream& paren,
+                                        ProofLetMap& letMap) = 0;
+
+  /**
+   * Print a proof of the empty clause.
+   *
+   * @param os the stream to print to
+   * @param paren any parentheses to add to the end of the global proof
+   */
+  virtual void printEmptyClauseProof(std::ostream& os, std::ostream& paren) = 0;
```
[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply src/proof/bitvector_proof.h]: Sounds good! Will do.

[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply src/proof/bitvector_proof.h]: Whoops! Yeah, it should be virtual. I'll fix that.

File: src/proof/bitvector_proof.h [hunk 4]
```diff
@@ -153,13 +181,40 @@ class BitVectorProof : public TheoryProof
...
+   * Kind of a mess. Used for resulution-based BVP's, where in eager mode this
+   * must be invoked before printing a proof of the empty clause. In lazy mode
+   * the behavior and purpose are both highly unclear.
+   *
+   * This exists as a virtual method of BitvectorProof, and not
+   * ResolutionBitvectorProof, because the machinery that invokes it is
+   * high-level enough that it doesn't know the difference between clausal and
+   * resolution proofs.
+   *
+   * TODO(aozdemir) figure out what is going on and clean this up
```
[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply src/proof/bitvector_proof.h]: Sounds good! Will do.

[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply src/proof/bitvector_proof.h]: [here](https://github.com/CVC4/CVC4/issues/2789)

File: src/proof/clausal_bitvector_proof.cpp (hunk 2 above)
Comment (alex-ozdemir) [reply src/proof/clausal_bitvector_proof.cpp]: Good idea!

File: src/proof/clausal_bitvector_proof.h [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+ **   Alex Ozdemir
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2018 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Bitvector proof using the DRAT & LRAT proof formats
+ **
+ ** An internal string stream is hooked up to cryptominisat, which spits out a
```
[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply src/proof/clausal_bitvector_proof.h]: Whoops, this whole paragraph is somewhat out of date. I'll update it.

File: src/proof/cnf_proof.cpp [hunk 6]
```diff
@@ -105,6 +105,30 @@ void CnfProof::registerConvertedClause(ClauseId clause, bool explanation) {
   setClauseDefinition(clause, current_expr);
 }
 
+void CnfProof::registerTrueUnitClause(ClauseId clauseId)
+{
+  Node true_node = NodeManager::currentNM()->mkConst<bool>(true);
```
[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply src/proof/cnf_proof.cpp]: Whoops! I'll fix it. I tend to accidentally use snake_case a lot.

File: src/proof/cnf_proof.cpp [hunk 7]
```diff
@@ -105,6 +105,30 @@ void CnfProof::registerConvertedClause(ClauseId clause, bool explanation) {
...
+  registerConvertedClause(clauseId);
+  popCurrentAssertion();
+  popCurrentDefinition();
+  d_cnfStream->ensureLiteral(true_node);
+  d_trueUnitClause = clauseId;
+}
+
+void CnfProof::registerFalseUnitClause(ClauseId clauseId)
+{
+  Node false_node = NodeManager::currentNM()->mkConst<bool>(false).notNode();
```
[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply src/proof/cnf_proof.cpp]: :+1:

File: src/proof/resolution_bitvector_proof.h [hunk 8]
```diff
@@ -106,23 +85,24 @@ class ResolutionBitVectorProof : public BitVectorProof
 
   bool d_isAssumptionConflict;
 
-  theory::TheoryId getTheoryId() override;
 };
 
-class LFSCBitVectorProof : public ResolutionBitVectorProof
+class LFSCResolutionBitVectorProof : public ResolutionBitVectorProof
```
[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply src/proof/resolution_bitvector_proof.h]: Good catch! I'll fix it!

File: src/proof/theory_proof.cpp [hunk 9]
```diff
@@ -105,10 +117,11 @@ void TheoryProofEngine::finishRegisterTheory(theory::Theory* th) {
   if (th) {
     theory::TheoryId id = th->getId();
     if (id == theory::THEORY_BV) {
+      theory::bv::TheoryBV* bv_th = static_cast<theory::bv::TheoryBV*>(th);
```
[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply src/proof/theory_proof.cpp]: Whoops! Thanks! I'll fix it.

File: src/prop/cnf_stream.cpp [hunk 10]
```diff
@@ -84,14 +84,13 @@ void CnfStream::assertClause(TNode node, SatClause& c) {
...
-       d_cnfProof->popCurrentDefinition();
-     }
-    );
+  PROOF(if (d_cnfProof) {
+    if (clause_id != ClauseIdError)
+    {
+      d_cnfProof->registerConvertedClause(clause_id);
+    }
+    d_cnfProof->popCurrentDefinition();
+  });
```
[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply src/prop/cnf_stream.cpp]: Sounds good!

[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply src/prop/cnf_stream.cpp]: That is much better!

File: src/prop/cryptominisat.cpp [hunk 11]
```diff
@@ -117,9 +118,17 @@ ClauseId CryptoMinisatSolver::addClause(SatClause& clause, bool removable){
   
   std::vector<CMSat::Lit> internal_clause;
   toInternalClause(clause, internal_clause);
-  bool res = d_solver->add_clause(internal_clause);
-  d_okay &= res;
-  return ClauseIdError;
+  bool nowOkay = d_solver->add_clause(internal_clause);
+  int freshId = ClauseId(ProofManager::currentPM()->nextId());
```
[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply src/prop/cryptominisat.cpp]: Whoops! Yeah.

File: src/theory/bv/bitblast/eager_bitblaster.cpp [hunk 12]
```diff
@@ -268,11 +268,12 @@ bool EagerBitblaster::collectModelInfo(TheoryModel* m, bool fullModel)
...
-void EagerBitblaster::setResolutionProofLog(
-    proof::ResolutionBitVectorProof* bvp)
+void EagerBitblaster::setProofLog(proof::BitVectorProof* bvp)
 {
-  THEORY_PROOF(d_bvp = bvp; d_satSolver->setProofLog(bvp);
-               bvp->initCnfProof(d_cnfStream.get(), d_nullContext.get());)
+  THEORY_PROOF(d_bvp = bvp; bvp->attachToSatSolver(*d_satSolver);
+               prop::SatVariable t = d_satSolver->trueVar();
+               prop::SatVariable f = d_satSolver->falseVar();
+               bvp->initCnfProof(d_cnfStream.get(), d_nullContext.get(), t, f));
```
[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply src/theory/bv/bitblast/eager_bitblaster.cpp]: Good spot! Yeah, I should have seen that.

File: src/theory/bv/bitblast/lazy_bitblaster.cpp [hunk 13]
```diff
@@ -566,11 +566,12 @@ bool TLazyBitblaster::collectModelInfo(TheoryModel* m, bool fullModel)
...
-void TLazyBitblaster::setProofLog(proof::ResolutionBitVectorProof* bvp)
+void TLazyBitblaster::setProofLog(proof::BitVectorProof* bvp)
 {
-  d_bvp = bvp;
-  d_satSolver->setProofLog( bvp );
-  bvp->initCnfProof(d_cnfStream.get(), d_nullContext.get());
+  THEORY_PROOF(d_bvp = bvp; bvp->attachToSatSolver(*d_satSolver);
+               prop::SatVariable t = d_satSolver->trueVar();
+               prop::SatVariable f = d_satSolver->falseVar();
+               bvp->initCnfProof(d_cnfStream.get(), d_nullContext.get(), t, f));
```
[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply src/theory/bv/bitblast/lazy_bitblaster.cpp]: Agreed!

**Comment by alex-ozdemir** [comment]
Thanks for the rapid review, Andres! I believe that I've addressed all of your comments.
If that's not the case, let me know. Also let me know if you think further work should be done to address any of them.

**Comment by 4tXJ7f** [comment]
Some regressions seem to be failing. Could you take a look?

**Comment by alex-ozdemir** [comment]
Yeah, sorry, I'll do so, and report back when they're fixed.

**Comment by alex-ozdemir** [comment]
Whoops -- I messed up the ordering of members within `TBitblaster`.

It turns out that the lazy and eager bitblasters ordered their (common) members differently. I hope that I have found an ordering that works for both!

**Comment by alex-ozdemir** [comment]
Okay, it seems that this new ordering (the one from the lazy bitblaster) works!

**Comment by alex-ozdemir** [comment]
ASAN reports some memory leaks. I don't _think_ they're related to this though.

**Comment by 4tXJ7f** [comment]
Sorry, I should have warned you. The leaks are a known issue. I actually just created a GitHub issue to track it (#2790). Removing the "do not merge" label.

**Pull request merged**